File: LeetCode/backtracking/77.combinations.py
#
# @lc app=leetcode id=77 lang=python3
#
# [77] Combinations
#

# @lc code=start
class Solution:
    result = []
    k = 0
    n = 0
    def combine(self, n: int, k: int) -> List[List[int]]:
        self.result = []
        self.k = k
        self.n = n
        self.backtrack([], 1)
        return self.result

    def backtrack(self, partialSol, startNum):
        if len(partialSol) == self.k:
            self.result.append(partialSol.copy())
            return 
        for num in range(startNum, self.n+1):
            partialSol.append(num)
            self.backtrack(partialSol, num+1)
            partialSol.pop(-1)

        
        
# @lc code=end

File: LeetCode_Go/Backtrack/77.combinations.go
/*
 * @lc app=leetcode id=77 lang=golang
 *
 * [77] Combinations
 */

// @lc code=start

// Given two integers n and k, return combinations of k numbers chosen from the range [1, n].

// all possible combinations of k numbers chosen from the range [1, n], essentially all subsets of [1, n] of size k (ref. 78.subsets). backtracking(start:) where we only branch from number whose is greater than start. choiceSet: all number greater than the current number

func combine(n int, k int) [][]int {
	ans := [][]int{}
	backtracking(1, n, k, []int{}, &ans)
	return ans
}

func backtracking(start, n, k int, comb []int, ans *[][]int) {
	// exit condition: when size is k, save to ans
	if len(comb) == k {
		temp := make([]int, k)
		copy(temp, comb)
		*ans = append(*ans, temp)
		return
	}

	for i := start; i <= n; i++ {
		// do choice
		comb = append(comb, i)
		backtracking(i+1, n, k, comb, ans)

		// undo choice
		comb = comb[:len(comb)-1]
	}
}

// @lc code=end

